src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs

```diff
@@ -507,40 +496,41 @@ private ReadOnlySpan<byte> InternalRead(int numBytes)
         // reasons. More about the subject in: https://github.com/dotnet/coreclr/pull/22102
         protected virtual void FillBuffer(int numBytes)
         {
-            if (numBytes < 0 || numBytes > _buffer.Length)
-            {
-                throw new ArgumentOutOfRangeException(nameof(numBytes), SR.ArgumentOutOfRange_BinaryReaderFillBuffer);
-            }
+            ArgumentOutOfRangeException.ThrowIfNegative(numBytes);
 
             ThrowIfDisposed();
 
-            // Need to find a good threshold for calling ReadByte() repeatedly
-            // vs. calling Read(byte[], int, int) for both buffered & unbuffered
-            // streams.
-            if (numBytes == 1)
-            {
-                int n = _stream.ReadByte();
-                if (n == -1)
-                {
-                    ThrowHelper.ThrowEndOfFileException();
-                }
-
-                _buffer[0] = (byte)n;
-                return;
-            }
-
-            if (numBytes > 0)
+            switch (numBytes)
             {
-                _stream.ReadExactly(_buffer.AsSpan(0, numBytes));
-            }
-            else
-            {
-                // ReadExactly no-ops for empty buffers, so special case numBytes == 0 to preserve existing behavior.
-                int n = _stream.Read(_buffer, 0, 0);
-                if (n == 0)
-                {
-                    ThrowHelper.ThrowEndOfFileException();
-                }
+                case 0:
+                    // ReadExactly no-ops for empty buffers, so special case numBytes == 0 to preserve existing behavior.
+                    int n = _stream.Read(Array.Empty<byte>(), 0, 0);
+                    if (n == 0)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    break;
+                case 1:
+                    // Need to find a good threshold for calling ReadByte() repeatedly
+                    // vs. calling Read(byte[], int, int) for both buffered & unbuffered
+                    // streams.
+                    n = _stream.ReadByte();
+                    if (n == -1)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    break;
+                default:
+                    byte[] buffer = ArrayPool<byte>.Shared.Rent(numBytes);
+                    try
+                    {
+                        _stream.ReadExactly(buffer.AsSpan(0, numBytes));
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(buffer);
+                    }
+                    break;
```

Comment:
The description of this method says:

> Fills the internal buffer with the specified number of bytes read from the stream.

These changes are breaking the public contract as they don't fill the internal buffer.

Our test coverage for it is far from ideal:

https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.IO/tests/BinaryReader/BinaryReaderTests.cs#L441-L462

I know that we should most likely not expose such method in the first place, as it limits the number of optimizations we can apply, but breaking it is imo not acceptable.

@stephentoub what is your opinion on this matter?

Reply:
This "internal buffer" that gets filled is otherwise unused by the rest of the code and inaccessible to the user. `FillBuffer` still satisfies its contract by reading from the stream.

Reply:
> what is your opinion on this matter?

Is there any way in which the change is visible via public API?  If the answer is no, I think it's fine.  We should also obsolete FillBuffer separately.

Reply:
> Is there any way in which the change is visible via public API? If the answer is no, I think it's fine.

Thansk, in such case I am fine with it too.